Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,710 @@
+  have {a b c : ℝ} : a - b - c = a + (-b - c) := by ring
+  simp_rw [this]
+  rw [deriv_add]
+  · rw [show log (q - 1) + (1 - x).log - x.log = log (q - 1) + ((1 - x).log - x.log) by
+      exact add_sub_assoc (log (q - 1)) (1 - x).log x.log]
+    congr! 1
+    simp only [differentiableAt_id', differentiableAt_const, deriv_mul, deriv_id'', one_mul,
+      deriv_const', mul_zero, add_zero]
+    convert deriv_binaryEntropy' using 2
+    simp only [neg_mul]
+  · simp only [differentiableAt_id', differentiableAt_const, DifferentiableAt.mul]
+  · apply DifferentiableAt.sub
+    · exact DifferentiableAt.neg
+        ((DifferentiableAt.mul differentiableAt_id') (DifferentiableAt.log differentiableAt_id' h))
+    · apply DifferentiableAt.mul (differentiable_const_minus x)
+        (DifferentiableAt.log (differentiable_const_minus x) (sub_ne_zero_of_ne hh.symm))
+
+lemma deriv_qaryEntropy {q : ℕ} {x : ℝ} (h: x ≠ 0) (hh : x ≠ 1) :
+    deriv (qaryEntropy q) x = log (q - 1) + log (1 - x) - log x := by
+  unfold qaryEntropy
+  exact deriv_qaryEntropy' h hh
+
+/-- Binary entropy has derivative `log (1 - p) - log p`.
+It's not differentiable at `0` or `1` but the junk values of `deriv` and `log` coincide there. -/
+lemma deriv_binaryEntropy {x : ℝ} :
+    deriv binaryEntropy x = log (1 - x) - log x := by
+  simp only [binaryEntropy_eq]
+  exact deriv_binaryEntropy'
+
+/-- Binary entropy has derivative `log (1 - p) - log p`. -/
+lemma hasDerivAt_binaryEntropy {x : ℝ} (xne0: x ≠ 0) (xne1 : x ≠ 1) :
+    HasDerivAt binaryEntropy (log (1 - x) - log x) x := by
+  convert hasDerivAt_deriv_iff.mpr (differentiableAt_binaryEntropy xne0 xne1) using 1
+  · exact binaryEntropy_eq
+  · exact (deriv_binaryEntropy').symm
+
+lemma hasDerivAt_qaryEntropy {q : ℕ} {x : ℝ} (xne0: x ≠ 0) (gne1 : x ≠ 1) :
+    HasDerivAt (qaryEntropy q) (log (q - 1) + log (1 - x) - log x) x := by
+  have diffAt :
+      DifferentiableAt ℝ (fun p => p * log (q - 1) - p * log p - (1 - p) * log (1 - p)) x := by
+    have : (fun p => p * log (q - 1) - p * log p - (1 - p) * log (1 - p)) =
+      (fun p => p * log (q - 1) + (-p * log p - (1 - p) * log (1 - p))) := by ext; ring
+    rw [this]
+    exact DifferentiableAt.add (by fun_prop) (differentiableAt_binaryEntropy xne0 gne1)
+  convert hasDerivAt_deriv_iff.mpr diffAt using 1
+  exact (deriv_qaryEntropy' xne0 gne1).symm
+
+open Filter Topology Set
+
+lemma eventuallyEq_nhdsWithin_of_eqOn_interval_right
+    (f g : ℝ → ℝ) (x ε : ℝ) (epsPos : 0 < ε) (h : ∀ y ∈ Ioo x (x + ε), f y = g y) :
+    f =ᶠ[𝓝[>] x] g := by
+  apply eventuallyEq_nhdsWithin_iff.mpr
+  apply Metric.eventually_nhds_iff.mpr
+  use ε
+  constructor
+  · exact epsPos
+  · intro y yclose ygex
+    have : y ∈ Ioo x (x + ε) := by
+      simp_all only [mem_Ioo, and_imp, mem_Ioi, true_and]
+      have : dist y x = y - x := by
+        simp_all only [Real.dist_eq, abs_eq_self, sub_nonneg, le_of_lt ygex]
+      linarith
+    simp_all only [mem_Ioo, and_imp, mem_Ioi, true_and]
```

Comment:
```suggestion
  have : x < x + ε := lt_add_of_pos_right x epsPos
  filter_upwards [Ioo_mem_nhdsWithin_Ioi' this] with y
  exact h y
```
And this is probably worth just inlining.

Reply:
I would have liked to have this lemma when starting out (and had a very cumbersome proof, as you saw).

I think quite a few people starting to learn Mathlib don't "think in filters" at first and it's nice to have these lemmas that one can find using Loogle and then see how it's proved.

Reply:
Yes, I've definitely been there as well! I don't think having these lemmas in mathlib is the way to go however: the main reason we use filters in the first place is to avoid having lemmas for every conceivable kinds of limits; if we start, it will basically never end. So I think it's best to leave this to tutorials, see e.g. Chapter 9 of Mathematics in Lean.

Reply:
Inlined.